Sources/X509/Extensions.swift

```diff
@@ -133,48 +132,104 @@ extension Certificate.Extensions: Hashable { }
-        // TODO(cory): enforce uniqueness
-        self._extensions.append(ext)
+    public mutating func append(_ extension: Certificate.Extension) throws {
+        if let oldExtension = self._extensions.first(where: { $0.oid == `extension`.oid }) {
+            throw CertificateError.duplicateOID(reason: "tried to append an extension for OID \(`extension`.oid) which is already present. Old extension: \(oldExtension) New extension: \(`extension`)")
+        } else {
+            self._extensions.append(`extension`)
+        }
     }
-
-    /// Append a sequence of new ``Certificate/Extension``s into this set of ``Certificate/Extensions-swift.struct``.
+    
+    /// Updates the ``Certificate/Extension`` stored in the dictionary for the ``Certificate/Extension/oid`` of the `extension`,
+    /// or appends `extension` if an ``Certificate/Extension`` with same  ``Certificate/Extension/oid`` does not exist.
     ///
-    /// - Parameter extensions: The sequence of new ``Certificate/Extension``s to insert.
+    /// - Parameter extension: The ``Certificate/Extension`` to update or append.
+    /// - Returns: The old ``Certificate/Extension`` that was replaced or `nil` if no ``Certificate/Extension`` with same ``Certificate/Extension/oid`` was present
     @inlinable
-    public mutating func append<Extensions: Sequence>(contentsOf extensions: Extensions) where Extensions.Element == Certificate.Extension {
-        // TODO(cory): enforce uniqueness
-        self._extensions.append(contentsOf: extensions)
+    @discardableResult
+    public mutating func update(_ extension: Certificate.Extension) -> Certificate.Extension? {
+        if let index = self._extensions.firstIndex(where: { $0.oid == `extension`.oid }) {
+            let oldExtension = self._extensions[index]
+            self._extensions[index] = `extension`
+            return oldExtension
+        } else {
+            self._extensions.append(`extension`)
+            return nil
+        }
+    }
+    
+    /// Updates the ``Certificate/Extension`` stored in the dictionary for the ``Certificate/Extension/oid`` of the `extension`,
+    /// or appends `extension` if an ``Certificate/Extension`` with same  ``Certificate/Extension/oid`` does not exist.
+    ///
+    /// - Parameters:
+    ///   - extension: The ``Certificate/Extension`` to update or append.
+    ///   - index: The index at which to insert the ``Certificate/Extension``, if it doesn't already exist.
+    ///
+    /// - Returns: A pair `(old, index)`, where `old` is the ``Certificate/Extension`` that was
+    ///    replaced, or `nil`no ``Certificate/Extension`` with same ``Certificate/Extension/oid`` was present, and `index`
+    ///    is the index corresponding to the updated (or inserted) ``Certificate/Extension``.
+    @inlinable
+    @discardableResult
+    public mutating func update(
+        _ extension: Certificate.Extension,
+        insertingAt index: Int
```

Comment:
The behaviour of this is a bit weird, do we really want it? We only sometimes respect the index, which seems unlikely to be right. Surely we should _always_ respect the index.

Reply:
I have mirrored the API from `OrderedDictionary` as this type is essentially a `SmallOrderedDictionary`: https://swiftpackageindex.com/apple/swift-collections/1.0.4/documentation/orderedcollections/ordereddictionary/updatevalue(_:forkey:insertingat:)

Reply:
Thinking about this more I think it would be weird to do always insert it at `index` because we would need to first remove an extension from `self ` with same `oid` which modifies the collection and kind of invalidates the `index`. You can only safely do it if you know beforehand if the element is present or not or otherwise the index might be outside the bounds.